### PR TITLE
[WIP] Add ScrapedPage class

### PR DESCRIPTION
**Pull request by chrismytton**
Encapsulate the details of making an HTTP request and wrapping the response body in a nokogiri instance behind a simple interface.

### CODE REVIEW

File: scraper.rb [hunk 1]
```diff
@@ -13,9 +13,22 @@
   2008 => 'List_of_MPs_elected_in_the_Mongolian_legislative_election,_2008',
 }
 
+class Interaction
+  def initialize(url)
+    @url = url
+  end
+
+  def noko
+    @noko ||= Nokogiri::HTML(open(url).read)
```
Comment (chrismytton) [inline scraper.rb]: I guess the `open-uri` response could probably be pulled into it's own method to make things a bit clearer.

File: scraper.rb [hunk 2]
```diff
@@ -13,9 +13,26 @@
   2008 => 'List_of_MPs_elected_in_the_Mongolian_legislative_election,_2008',
 }
 
+class Interaction
+  def initialize(url)
+    @url = url
+  end
+
+  def noko
+    @noko ||= Nokogiri::HTML(response.read)
```
Comment (chrismytton) [inline scraper.rb]: The `response.read` here will mean that anything else that reads the response will need to remember to call `response.rewind` first. So probably best to pull the response body into it's own method that handles that.

**Comment by chrismytton** [comment]
@tmtmtmtm 👀

File: lib/scraped_page.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
+class ScrapedPage
+  def initialize(url)
+    @url = url
+  end
+
+  def noko
+    @noko ||= Nokogiri::HTML(response_body)
+  end
```
Comment (chrismytton) [inline lib/scraped_page.rb]: This can probably be private now it's not being used from the outside.

File: lib/scraped_page.rb [hunk 4]
```diff
@@ -0,0 +1,32 @@
+require 'field_serializer'
+require 'nokogiri'
+require 'open-uri'
+
+class ScrapedPage
+  include FieldSerializer
```
Comment (chrismytton) [inline lib/scraped_page.rb]: I'm not sure having this included in `ScrapedPage` makes sense actually. Perhaps we should encourage the user to subclass `ScrapedPage` as `Page` and then include `FieldSerializer` in that?

**Comment by chrismytton** [comment]
I've changed things so that there's now a `ScrapedPage` class which provides the encapsulation of requests. This makes sense (to me at least!) because a term page _is_ a scraped page.

If the user wants to perform further customization then they can subclass `ScrapedPage` themselves:

``` ruby
class Page < ScrapedPage
  # Archive related bits
  archive_url ENV['MORPH_ARCHIVE_URL']
  archive_refresh :always
  request_timeout 30
  request_retries 3
end

class HistoricTermPage < Page
  # Only update the archive once a week
  archive_refresh :weekly
end
```

**Comment by chrismytton** [comment]
Perhaps `Page` wasn't the best example class name there, it would make sense for it to be something more specific to the scraper, `MongoliaKhuraiPage` or similar.

File: lib/scraped_page.rb (hunk 4 above)
Comment (chrismytton) [reply lib/scraped_page.rb]: I take it back, if `ScrapedPage` is going to be an abstract base class that all scrapers use then it totally makes sense for it to include `FieldSerializer`.

**Comment by tmtmtmtm** [comment]
@chrismytton has this been obsoleted by #12?

**Comment by chrismytton** [comment]
> @chrismytton has this been obsoleted by #12?

Indeed it has!

**Pull request closed without merging**